File: spark/dataframes/corruptRecords.scala
import org.apache.spark.sql.types._

// Put the json files in data subdirectory into HDFS prior to running this
// hdfs dfs -put hadoop-examples-data
val maxValsData = "hadoop-examples-data/maxVals.json"

// Run this code using spark-shell using :load <this file>
// (recommend using `:silent` prior to calling
// to silence the output of defining the functions)

// Comment / uncomment the following function calls to see 
// how spark's dataframes work w/schemas
useNoSchema()

useStringSchema()

useIntegerSchemaNoCorruptRecord()

// Cause crash due to 
//   https://issues.apache.org/jira/browse/SPARK-29621
//   and https://issues.apache.org/jira/browse/SPARK-21610
useIntegerSchemaAndCorruptRecord(showCount = true)

// Avoid crash
useIntegerSchemaAndCorruptRecord(showCount = false)

// Use different error column name
useRufus()
useSchemaWithLongValue()

println("Done")
def useNoSchema() {
    val jsonDF = spark.read.json(maxValsData)
    println("This is with no schema:")
    jsonDF.printSchema()
}

def useStringSchema() {
    val columnsList = List(
	    StructField("product", StringType),
	    StructField("price", StringType)
    )
    val jsonWithSchema = spark.read.schema(StructType(columnsList)).json(maxValsData)
    println("This is with String Schema")
    jsonWithSchema.show()
}


// ------ Try specifying a schema with string and INT
// ------ And with no _corrupt_record field
def useIntegerSchemaNoCorruptRecord() {
    val columnsList = List(
        StructField("product", StringType),
        StructField("price", IntegerType)
    )
    val jsonWithBadSchema = spark.read.schema(StructType(columnsList)).json(maxValsData)
    println("This is with String and INT Schema, but no _corrupt_record field")
    jsonWithBadSchema.show()
}


// ------ Specify schema with _corrupt_record and try to count()
// Using .count() with _corrupt_record will crash:
//   See: https://issues.apache.org/jira/browse/SPARK-29621
//   and https://issues.apache.org/jira/browse/SPARK-21610
// You can change the name of _corrupt_record field, see "useRufus" below
def useIntegerSchemaAndCorruptRecord(showCount: Boolean = false) {
    val columnsList = List(
        StructField("product", StringType),
        StructField("price", IntegerType),
        StructField("_corrupt_record", StringType)
    )

    val jsonWithBadSchemaAndCorruptRecordField = spark.read.schema(StructType(columnsList)).json(maxValsData)
    val errors = jsonWithBadSchemaAndCorruptRecordField.filter("_corrupt_record IS NULL")
    if (showCount) {
        println("This is using _corrupt_record, but count doesn't work:")
        errors.count()
    } else {
        println("This is using _corrupt_record show *DOES* work :-O")
        errors.show()
   }
}

// Change 
// Make sure to change _corrupt_record field's name in schema to "rufus"
def useRufus() {
    val columnsList = List(
        StructField("product", StringType),
        StructField("price", IntegerType),
        StructField("rufus", StringType)
    )

    val jsonWithDifferentErrorColumn = spark.read.schema(StructType(columnsList))
        .option("columnNameOfCorruptRecord", "rufus")
        .json(maxValsData)
    println("This is with error column = \"rufus\"")
    jsonWithDifferentErrorColumn.show()
}

def useSchemaWithLongValue() {
    // ----- Try using a schema with Long and it works :-)
    val columnsList = List(
        StructField("product", StringType),
        StructField("price", LongType),
        StructField("_corrupt_record", StringType)
    )
    val jsonWithLongSchema = spark.read.schema(StructType(columnsList)).json(maxValsData)
    println("This is json with LONG in Schema")
    jsonWithLongSchema.show()
}

File: spark/dataframes/corruptRecords.py
# Put data into HDFS prior to running
# hdfs dfs -put hadoop-examples-data
# For config options, see
#     https://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.DataFrameReader.json
from pyspark.sql.types import *
maxValsData = "hadoop-examples-data/maxVals.json"

jsonDF = spark.read.json(maxValsData)
jsonDF.printSchema()

# Try specifying a nice schema and see what happens
schema = StructType([
    StructField("product", StringType()),
    StructField("price", StringType())])

jsonWithNiceSchema = spark.read.schema(schema).json(maxValsData)
print("This is with schema of strings")
jsonWithNiceSchema.show()

# Specify a Schema with INT instead of Long and watch the fireworks
columns = [
    StructField("product", StringType()),
    StructField("price", IntegerType())
]
badSchema = StructType(columns)

jsonWithBadSchema = spark.read.schema(badSchema).json(maxValsData)
print("This is with no error handling")
jsonWithBadSchema.show()

# Specify a schema with _corrupt_record and Spark puts the bad records into
# _corrupt_record by default.  The name of this field can be configured by
# overriding spark.sql.columnNameOfCorruptRecord

# NOTE NOTE NOTE NOTE NOTE There's a bug attached with this capability,
# The problem is you need to cache() the data or write it to HDFS in
# order to filter on _corrupt_record :(
# https://issues.apache.org/jira/browse/SPARK-21610

columns = [ StructField("product", StringType()), 
    StructField("price", IntegerType()),
    StructField("_corrupt_record", StringType()) 
]

handleErrorsSchema = StructType(columns)
jsonWithHandleErrorsSchema = spark.read.schema(handleErrorsSchema).json(maxValsData)
print("This is with handling errors")
jsonWithHandleErrorsSchema.show()

# Override column name of bad record
columns = [ StructField("product", StringType()), 
    StructField("price", IntegerType()),
    StructField("rufus", StringType()) 
]
handleErrorsSchema = StructType(columns)
jsonWithDifferentErrorColumn = spark.read.schema(handleErrorsSchema)\
    .json(maxValsData, columnNameOfCorruptRecord = "rufus")

print("This is with handling errors in rufus column")
# jsonWithDifferentErrorColumn.show()
# This is required :-(
# jsonWithDifferentErrorColumn.unpersist()
print("These are good records:")
onlyGood = jsonWithDifferentErrorColumn.filter("rufus IS NULL")
onlyGood.show()
print("These are bad records:")
onlyBad = jsonWithDifferentErrorColumn.filter("rufus IS NOT NULL")
onlyBad.show()

